### PR TITLE
fix(slice): translate ONNX reverse slices to Burn range semantics

### LINKED ISSUE #476: Slice shape inference panics on i64::MIN sentinel (overflow in normalize_index's idx+n)
## Summary

`calculate_dim_slice_output_len()` in `crates/onnx-ir/src/node/slice.rs` normalizes negative indices with `idx + n` using plain addition:

```rust
fn normalize_index(idx: i64, lo: i64, hi: i64, n: i64) -> i64 {
    let abs = if idx < 0 { idx + n } else { idx };
    abs.clamp(lo, hi)
}
```

The ONNX Slice spec defines **`i64::MIN` as an explicit sentinel** (for reverse slicing, `end = INT64_MIN` means "before the first element"). When such a sentinel reaches this helper, `idx + n` overflows `i64`: debug builds **panic** ("attempt to add with overflow") during model import/shape inference, while release builds silently wrap and only then clamp. The function's own doc comment advertises exactly this sentinel handling ("for `step < 0`, `i64::MIN + dim_len` clamps to `-1`"), but the arithmetic panics before the clamp can run in debug configurations.

Static-analysis finding against current `master`; verified by code reading only.

## Location

- File: `crates/onnx-ir/src/node/slice.rs`
- Functions:
  - `normalize_index()` nested inside `calculate_dim_slice_output_len()` (~lines 88-91),
  - called from `calculate_dim_slice_output_len()` (~lines 76-107), which runs at node-processing time from the Slice processor (~lines 204 and ~247).

## Problem

For a reverse slice with `ends[0] == i64::MIN` and any tensor rank:

- `step < 0` branch → `normalize_index(i64::MIN, -1, dim_len - 1, dim_len)`
- `idx < 0` → evaluates `i64::MIN + dim_len`

That add is undefined-range arithmetic in Rust's terms: debug profile aborts with an overflow panic; release profile wraps to a large negative that then clamps to `-1`. The same applies to a (spec-legal but unusual) `starts` value of `i64::MIN` on the positive-step path.

Since this code executes at import/shape-inference time inside burn-onnx, importing a perfectly valid ONNX model - one that follows the documented `INT64_MIN` convention used by PyTorch/TensorFlow exporters for reverse slices - terminates with a panic in debug builds instead of computing the output shape.

## Trigger / Reproduction

Based on static analysis; no runtime run performed:

```text
Slice node attributes/inputs:
  starts = [dim_len - 1]        (e.g. [4])
  ends   = [-9223372036854775808]
  steps  = [-1]
```

Importing this model with burn-onnx hits `normalize_index` with `idx = i64::MIN`.

## Expected Behavior

Sentinel values must be handled with saturating/checked arithmetic so the documented clamping occurs identically in all build profiles:

```rust
let abs = if idx < 0 { idx.saturating_add(n) } else { idx };
```

(or compare before adding).

## Actual Behavior

Debug builds panic with an integer-overflow abort; release builds work only incidentally via wrap+clamp, making behavior configuration-dependent.

## Impact

Any user importing a real-world model whose exporter emitted the standard `INT64_MIN` reverse-slice sentinel gets a hard panic during code generation when building with `debug_assertions` (common for local development of Burn projects). Even where it "works", relying on wrap-around is fragile.

## Suggested Direction

Switch the single addition to `saturating_add` (and audit sibling helpers in this file for the same pattern, e.g. anywhere `starts`/`ends`/`axes` are offset by rank or dim length). A unit test feeding `(i64::MIN, x, -1, n)` through `calculate_dim_slice_output_len` would pin the fix; the file header already documents the sentinel semantics to test against.

## Evidence

- Arithmetic quoted above; Rust overflow semantics (panic under `debug_assertions`, two's-complement wrap otherwise).
- ONNX Slice documentation: negative `end` values below `-rank` and the `INT64_MIN` convention represent "clamp to boundary"/"before first element".

### PR DESCRIPTION
**Pull request by antimora**
### Checklist

- [x] Confirmed that `cargo xtask validate` command has been executed.
- [x] Confirm relevant documentation has been updated.
- [x] For new/modified ONNX operators: verified implementation matches `onnx-spec/ops/Slice.md`.

### Related Issues/PRs

Closes #476.

The overflow that issue reports is not reachable: `idx + n` there is `i64::MIN + dim_len` with `dim_len >= 0`, which moves toward zero. The `onnx-ir` slice suite passes under `RUSTFLAGS="-C overflow-checks=on"`, and `test_tensor_static_shape_end_i64_min_slices_to_start` already covered that exact case. But building the reporter's model surfaced a real bug in the same place, plus a genuine overflow elsewhere in the same function. Details in [this comment](https://github.com/tracel-ai/burn-onnx/issues/476#issuecomment-5400831164).

### Changes

**ONNX and Burn disagree about what a negative step means, and codegen was passing the values straight through.**

| | `start=0, end=5, step=-1` over a 5-element axis |
|---|---|
| Burn `s![0..5;-1]` | `[4, 3, 2, 1, 0]` — forward range, step sign reverses traversal |
| ONNX | `[]` — start at 0, walk backwards, stop before 5 |

So every negative-step slice was wrong, not just sentinel ones, and the ONNX `INT64_MIN` "past the first element" sentinel reached the generated source verbatim:

```rust
let slice1_out1 = x.slice(s![4.. - 9223372036854775808; - 1, ..]);
//  error: literal out of range for `i32`
```

A valid model that PyTorch-style exporters emit for `x[::-1]` produced a crate that would not compile.

`onnx-ir` had the output shape right all along; only codegen disagreed with it. The existing `slice_with_steps` test encoded the bug — it asserted `[5, 3, 12]` where `onnx.reference.ReferenceEvaluator` returns `[5, 3, 0]` for that model.

The mapping is `lo = norm_end + 1`, `hi = norm_start + 1` after resolving both bounds against the dimension size, which also turns the sentinel into `lo = 0` for free. Where the axis size is not static, the bounds are resolved from the tensor's own `dims()` at runtime. The Shape-input path had the same defect, where the element count disagreed with the rank `onnx-ir` inferred and the generated `try_into` would have panicked.

Also fixed, found while verifying:

- **A reverse `start` below `-dim` returned one element instead of none.** It means "before the beginning". numpy and `ReferenceEvaluator` both give empty; the spec prose saying starts clamp into `[0, dim-1]` is what misled the original code. Both crates shared the misreading, so they agreed with each other and shape inference never contradicted codegen.
- **`i64` overflow on extreme `step` values** in `calculate_dim_slice_output_len`: `step.abs()` panics on `i64::MIN` and `range_len + step.abs() - 1` overflows on `i64::MAX`, both under `overflow-checks`. `steps` is unchecked model input. Now `(range_len as u64).div_ceil(step.unsigned_abs())`.
- **A non-unit `steps` with runtime bounds was silently ignored** on the tensor path. The guard for this existed only in the `Shape` branch of `infer_types`; the `Tensor` branch had no step validation at all, so a model with runtime bounds and a constant `step = -1` imported cleanly and emitted a plain forward range. It now fails at import naming the node and the step. Runtime `steps` alongside runtime bounds stays supported — the backend suite's `test_slice` passes all params as graph inputs.

`slice_with_steps` was mis-specified rather than merely mis-asserted (`starts=0, ends=12, step=-1` selects nothing under ONNX), so the fixture now expresses the reverse its own comments describe. **Every original value assertion in that test holds unchanged.**

### Testing

Behaviour is verified against `onnx.reference.ReferenceEvaluator` throughout, not against the other implementation — the below-`-dim` bug is precisely the kind that cross-checking two same-authored implementations cannot find.

New integration models, each with its expectations taken from the reference evaluator:

- `slice_min_sentinel` — the `INT64_MIN` sentinel, two axes reversed at once
- `slice_reverse_steps` — `|step| > 1`, including a range length that is not a multiple of the step (the case a misaligned stride would expose), plus the reverse ONNX evaluates to nothing
- `slice_shape_reverse` — Shape-input reverse at `-1` and `-2`; this path sizes an array from `onnx-ir`'s inference, so a bounds disagreement panics in generated code rather than returning a wrong value
- `slice_reverse_dynamic` — a symbolic axis run at three input sizes with both a sentinel and a bounded `ends` at step `-2`, proving nothing is baked in at codegen time

Plus unit tests for `slice_bounds` and `calculate_dim_slice_output_len` across both step signs, both sentinels, negative and out-of-range bounds, and `dim == 0`; snapshot coverage for each new codegen branch; and `onnx-ir` guard tests including the all-runtime conformance shape.

`cargo xtask validate`: format, lint, typos and audit clean; 1191 `onnx-ir` (also under forced `overflow-checks`), 922 `burn-onnx`, 605 `onnx-tests` all pass. The workspace-test stage still fails to compile `ssmarshal`, a third-party `no_std` dependency of `examples/raspberry-pi-pico`; that reproduces on a clean tree at `main` and is unrelated to these changes.

### Note for reviewers

This changes results for any model using negative-step slices, so it is a correctness fix rather than a source-compatible one and may deserve a changelog entry.

Three pre-existing silent failures in this file are reported but deliberately **not** fixed here, since each needs a new `ProcessError` that would reject models importing today: a runtime `axes` input falls back to the leading dimensions, out-of-range and short `axes` entries are dropped, and the `i32` bound guard checks only `end`. Two already carry TODOs in `onnx-ir`. Happy to take any of them in a follow-up.

### CODE REVIEW

File: crates/burn-onnx/src/import/burn/node/slice.rs [hunk 1]
```diff
@@ -512,6 +485,142 @@ fn generate_tensor_slice(
...
+    }
+
+    // i64::MIN clamps to 0, so a forward step can never select anything. This
+    // has to be intercepted rather than passed through: Burn resolves a negative
+    // bound as `size as isize + index`, which overflows on i64::MIN.
+    if end == i64::MIN {
+        return Some(quote! { 0..0 });
+    }
+
+    let start = start.to_tokens();
```
[the remark this answers is not in the document]
Comment (antimora) [reply crates/burn-onnx/src/import/burn/node/slice.rs]: Good catch, and confirmed. I fixed the `end` sentinel and left the symmetric `start` case. Reproduced before fixing:

```rust
let y = x.slice(s![- 9223372036854775808..3]);
//  error: literal out of range for `i32`
```

Fixed in c9bd2d2. `start == i64::MIN` now emits `0`, which is what it clamps to for any dimension size, so `s![0..3]`. You're right about the overflow too, not just the `i32` inference: passing it through would have Burn evaluate `size as isize + isize::MIN` in `convert_signed_index`. That's why it has to be intercepted rather than clamped downstream, and the doc comment now says so.

The reverse path was already correct here, since it routes through `slice_bounds`, where the sentinel resolves via `saturating_add` and clamps to -1. Added `test_slice_forward_start_min_sentinel` to pin it.

File: crates/burn-onnx/src/import/burn/node/slice.rs [hunk 2]
```diff
@@ -512,6 +485,142 @@ fn generate_tensor_slice(
...
+    // i64::MAX means "to the end". The bound below it is emitted as a bare
+    // literal, so an absurd `end` is rejected here rather than becoming
+    // generated code that does not compile.
+    let end = if end == i64::MAX {
+        quote! {}
+    } else if end > i32::MAX as i64 {
+        panic!("Slice end index {} exceeds i32::MAX", end);
+    } else {
```
[the remark this answers is not in the document]
Comment (antimora) [reply crates/burn-onnx/src/import/burn/node/slice.rs:531]: Agreed the panic should go. I went a step further than moving it to `onnx-ir`, because on inspection the model does not need rejecting at all.

The bound only had to fit in `i32` because bare literals in an `s![..]` range infer as `i32`. Nothing downstream imposes that limit: Burn stores slice bounds as `isize`. Suffixing the literals that do not fit makes them compile, and Burn then clamps them exactly as the ONNX spec requires for out-of-range bounds. Verified against the backend:

```rust
s![0i64..3000000000i64]   // -> all 5 elements, end clamped to size
s![-3000000000i64..5i64]  // -> all 5 elements, start clamped to 0
```

So in c9bd2d2 the panic is gone and no model is rejected, which is better than converting it to a `ProcessError`: `ends = [3000000000]` legitimately means "to the end" and should just work.

In-range bounds still emit bare, so there is no snapshot churn; all 925 codegen tests pass unchanged. Added `test_slice_forward_bound_beyond_i32` for the suffixed path.

File: crates/onnx-ir/src/node/slice.rs [hunk 3]
```diff
@@ -182,6 +185,36 @@ impl NodeProcessor for SliceProcessor {
...
+                // runtime-bound paths emit a plain forward range, so a non-unit
+                // step there would silently slice the wrong elements. Reject the
+                // combination rather than ignoring the step, mirroring the guard
+                // the Shape input path already applies below.
+                //
+                // A runtime `steps` alongside runtime bounds stays allowed: the
+                // ONNX backend test suite passes all four as graph inputs, and
+                // the value cannot be checked here. Those models are assumed to
+                // step by 1, which is what every exporter emits.
+                let runtime_bounds = matches!(config.starts, SliceInput::Runtime(_))
```
[the remark this answers is not in the document]
Comment (antimora) [reply crates/onnx-ir/src/node/slice.rs:198]: Valid, and I took your second option. Rejecting is not available here: I tried exactly that first and `cargo xtask validate` caught it, because the ONNX backend test suite's `test_slice` passes `x`, `starts`, `ends`, `axes` and `steps` all as graph inputs with no initializers. A runtime `steps` is normal and valid there, so rejecting it breaks a conformance model.

c9bd2d2 emits the assertion instead, at the one point where the value exists:

```rust
assert!(
    step_vals.iter().all(|&s| s == 1),
    "Slice node {}: only step=1 is supported when `steps` is a runtime input",
    "slice1"
);
```

It handles `Shape`, `Tensor` and scalar `steps` arguments, and is emitted only when `steps` is runtime, so nothing else changes. `onnx-ir` already rejects a runtime `steps` alongside *static* bounds (that combination takes the static codegen path, which needs the values), so only the runtime-bound paths reach the assertion.

A silent wrong answer becomes a loud failure, and `test_slice` still imports and passes. 834 `onnx-official-tests` green.
